File: python/best_time_to_buy_and_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        lowest_buying_price = prices[0]
        max_profit = 0

        for price in prices:
            if price < lowest_buying_price:
                lowest_buying_price = price
            current_profit = price - lowest_buying_price
            if current_profit > max_profit:
                max_profit = current_profit
        return max_profit


trial = Solution()
stock_prices = [7, 6, 4, 3, 1, 9, 3, 7, 9, 1, 4, 6, 10]
print(trial.maxProfit(prices=stock_prices))

File: javascript/best_time_to_buy_and_sell_stock.js
function maximumProfit(prices) {
    lowest_buying_price = prices[0]
    max_profit = 0

    //  Note: Both methods work great. Try them out by uncommenting
    // prices.forEach(price => {
    //     if (price < lowest_buying_price) {
    //         lowest_buying_price = price
    //     }
    //     current_profit = price - lowest_buying_price
    //     if (current_profit > max_profit) {
    //         max_profit = current_profit
    //     }
    // }
    // )

    // for (let index = 0; index < prices.length; index++) {
    //     const price = prices[index];
    //     if (price < lowest_buying_price) {
    //         lowest_buying_price = price
    //     }
    //     current_profit = price - lowest_buying_price
    //     if (current_profit > max_profit) {
    //         max_profit = current_profit
    //     }
    // }

}

prices = [7, 6, 4, 3, 1, 9, 3, 7, 9, 1, 4, 6, 10]

console.log(maximumProfit(prices))